subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java

```diff
@@ -164,41 +163,67 @@ public DefaultJvmTestSuite(String name, DependencyFactory dependencyFactory, Con
-            // and add dependencies automatically.
-            getVersionedTestingFramework().convention((VersionedTestingFramework) null);
         }
 
         addDefaultTestTarget();
 
-        // Until the values here can be finalized upon the user setting them (see the org.gradle.api.tasks.testing.Test#testFramework(Closure) method),
-        // in Gradle 8, we will be executing the provider lambda used as the convention multiple times.  So make sure, within a Test Suite, that we
-        // always return the same one via computeIfAbsent() against this map.
-        final Map<TestingFramework, TestFramework> frameworkLookup = new HashMap<>(4);
-
         this.targets.withType(JvmTestSuiteTarget.class).configureEach(target -> {
             target.getTestTask().configure(task -> {
-                task.getTestFrameworkProperty().convention(getVersionedTestingFramework().map(vtf -> {
-                    switch(vtf.getFramework()) {
-                        case JUNIT4:
-                            return frameworkLookup.computeIfAbsent(vtf.getFramework(), f -> new JUnitTestFramework(task, (DefaultTestFilter) task.getFilter(), false));
-                        case KOTLIN_TEST: // fall-through
-                        case JUNIT_JUPITER: // fall-through
-                        case JUNIT_PLATFORM: // fall-through
-                        case SPOCK:
-                            return frameworkLookup.computeIfAbsent(vtf.getFramework(), f -> new JUnitPlatformTestFramework((DefaultTestFilter) task.getFilter(), false));
-                        case TESTNG:
-                            return frameworkLookup.computeIfAbsent(vtf.getFramework(), f -> new TestNGTestFramework(task, task.getClasspath(), (DefaultTestFilter) task.getFilter(), getObjectFactory()));
-                        default:
-                            throw new IllegalStateException("do not know how to handle " + vtf);
+
+                // By default, our versioned testing framework is whatever the Test says their framework is.
+                // However, `setFrameworkTo` will override out versioned testing framework with a `set` call and
+                // further `set` the Test's test framework.
+                // From now on, the test suite is "in control" of the framework. Before then, the test had complete
+                // control over which framework it is using. This allows the users to continue to interact with
+                // Test tasks directly and the test suite handles those cases.
+                // Regardless of how our versioned testing framework is calculated (either by convention by reading the
+                // Test's test framework or via a `setFrameworkTo` call), we will add the proper dependencies to the
+                // Test's corresponding source set.
+                getVersionedTestingFramework().convention(task.getTestFrameworkProperty().map(framework -> {
```

Comment:
I think we talked about this a bit on Friday, but this is backwards from the way we want this to work. The test task should be told what to do, so we shouldn't be looking at the test task to figure out which dependencies are needed. 

There are three types of `Test` tasks:
1. Tasks created without a test suite. For those, we don't do anything special.
2. Tasks created by a test suite not named "test". Here, we can be more restrictive and disallow changes to the Test task that conflict with the test suite.
3. Task created by a test suite named "test".  We can be a little restrictive, but mostly we need to keep existing builds working and then nag to move configuration into test suites (once it's stabilized).

Reply:
I do agree it is backwards. My main driving goal here is to stop us from needing to load test framework dependencies from the gradle distribution. We are still doing that for any test where the user directly interacts with the test to set the framework (which is right now the only supported way). 

This solution, while not really what we want in the long run, supports users interacting with tests directly for the time being -- while test suites are still incubating. Perhaps we can wait a bit on fixing this, though if thats the case I'd be curious if we have a timeline on when we actually plan to deincubate test suites -- or at least the portion that users interact with to test frameworks (I know a lot of this rides on the changes we've all been discussing recently in terms of messing around with how test frameworks are actually implemented). 

Another alternative solution here would be if the user calls `test.useXXX`, the test could lookup what suite its in and call its suite's corresponding `useXXX` method (which then sets the proper versioned test framework and dependencies), but this solution requires tests to be able to "see" a test suite, which they cannot right now due to their subprojects (something my other PR would resolve, unblocking this alternative solution). 

The end goal is to detect cases where a test relies on modules from the gradle distribution, but this would throw a lot of warnings for users calling `test.useXXX` methods, so we need to fix that first before adding a new deprecation warning.